### PR TITLE
Rover: Loiter in AUTO mode for boat

### PR DESCRIPTION
**Pull request by tajisoft**
I’m still in test. But I want someone to review this. Because I change to use rover.mode_rtl instead of _mode_rtl.

### CODE REVIEW

**Comment by rmackay9** [comment]
Yes, I think it's OK to reach up into the rover code and grab the loiter and rtl objects instead of passing them in.  I wish we could separate the flight mode class more from rover but there are still a lot of links so this is probably OK.

**Comment by tajisoft** [comment]
@rmackay9 Thank you for your feedback. I understood the implementation policy.

File: APMrover2/mode_auto.cpp [hunk 1]
```diff
@@ -50,13 +44,16 @@ void ModeAuto::update()
...
                 // continue driving towards destination
                 calc_steering_to_waypoint(_reached_destination ? rover.current_loc : _origin, _destination, _reversed);
                 calc_throttle(calc_reduced_speed_for_turn_or_distance(_reversed ? -_desired_speed : _desired_speed), true, true);
             } else {
-                // we have reached the destination so stop
-                stop_vehicle();
+                if (rover.is_boat() && !start_loiter()) {
+                    stop_vehicle();
+                } else {
+                    stop_vehicle();
```
Comment (magicrub) [inline APMrover2/mode_auto.cpp]: This logic is weird, If all you're trying to do is set Auto_Loiter then do it here or abstract this whole logic into the start_loiter() function. As-is, stop_vehicle() is called no matter what so this whole thing turns into:
```
if (rover.is_boat()) {
  start_loiter();
}
stop_vehicle();
```

File: APMrover2/mode_auto.cpp [hunk 2]
```diff
@@ -70,13 +67,21 @@ void ModeAuto::update()
                 // check if we have reached within 5 degrees of target
                 _reached_heading = (fabsf(_desired_yaw_cd - ahrs.yaw_sensor) < 500);
             } else {
-                stop_vehicle();
+                if (rover.is_boat() && !start_loiter()) {
+                    stop_vehicle();
+                } else {
+                    stop_vehicle();
```
Comment (magicrub) [inline APMrover2/mode_auto.cpp]: same

**Comment by magicrub** [comment]
try again

File: APMrover2/mode_auto.cpp (hunk 1 above)
Comment (tajisoft) [reply APMrover2/mode_auto.cpp]: @magicrub Thank you for your feedback! I don't want to call stop_vehicle() when start_loiter() returns true.
(rover.is_boat() && !start_loiter()) || stop_vehicle();
Should I not write like this? If so, I will fix it as you pointed out.

**Comment by tajisoft** [comment]
@magicrub Thank you for your advice!

**Comment by magicrub** [comment]
> I don't want to call stop_vehicle() when start_loiter() returns true.

@tajisoft ok, I see. Well your recent change doesn't do that either. How about this:
```
// we have reached the destination so stay here
if (rover.is_boat() && start_loiter()) {
  break;
}
stop_vehicle();
```

It could also get rolled into the above else that way

**Comment by rmackay9** [comment]
Here's this PR with the "strange logic" thing modified to be wordier but easier to understand I think.
https://github.com/rmackay9/rmackay9-ardupilot/commit/f414845dadad3e94850efeee0ecaeb8cb7517791

**Comment by rmackay9** [comment]
Merged with the change mentioned above.  Thanks!

**Pull request closed without merging**